keepassxc-browser/background/client.js

```diff
@@ -51,78 +51,86 @@ const kpErrors = {
+}
+
 //--------------------------------------------------------------------------
 // Messaging
 //--------------------------------------------------------------------------
 
-keepassClient.sendNativeMessage = function(request, enableTimeout = false, timeoutValue) {
-    return new Promise((resolve, reject) => {
-        let timeout;
-        const requestAction = request.action;
-        const ev = keepassClient.nativePort.onMessage;
-
-        const listener = ((port, action) => {
-            const handler = (msg) => {
-                if (msg && msg?.action === action) {
-                    // If the request has a separate requestID, check if it matches when there's no nonce (an error message)
-                    const isNotificationOrError = !msg.nonce && (request.requestID === msg.requestID || (msg.error && msg.errorCode));
-
-                    // Only resolve a matching response or a notification (without nonce)
-                    if (isNotificationOrError || messageBuffer.matchAndRemove(msg)) {
-                        port.removeListener(handler);
-                        if (enableTimeout) {
-                            clearTimeout(timeout);
-                        }
-
-                        resolve(msg);
-                        return;
-                    }
-                }
-            };
-            return handler;
-        })(ev, requestAction);
-        ev.addListener(listener);
-
-        const messageTimeout = timeoutValue || keepassClient.messageTimeout;
-
-        // Handle timeouts
-        if (enableTimeout) {
-            timeout = setTimeout(() => {
-                const errorMessage = {
-                    action: requestAction,
-                    error: kpErrors.getError(kpErrors.TIMEOUT_OR_NOT_CONNECTED),
-                    errorCode: kpErrors.TIMEOUT_OR_NOT_CONNECTED
-                };
-                keepass.isKeePassXCAvailable = false;
-                ev.removeListener(listener.handler);
-                resolve(errorMessage);
-            }, messageTimeout);
-        }
+keepassClient.sendNativeMessage = async function(request, enableTimeout = false, timeoutValue) {
+    if (!keepassClient.nativePort) {
+        return;
+    }
 
-        // Store the request to the buffer
-        messageBuffer.addMessage(request);
+    const message = new Message(request, enableTimeout, timeoutValue);
+    await navigator.locks.request('messageBuffer', async (lock) => {
+        messageBuffer.addMessage({ request: request, message: message });
+    });
+
+    keepassClient.nativePort.postMessage(request);
+    return await message.promise;
+};
+
+keepassClient.handleNativeMessage = async function(response) {
+    const isError = Boolean(!response.nonce && response.error && response.errorCode);
+
+    // Parse through the message buffer to find the corresponding Promise.
+    await navigator.locks.request('messageBuffer', async (lock) => {
+        for (let i = 0; i < messageBuffer.buffer.length; ++i) {
```

Comment:
Could use a logged error if there is no matching message found (ie, the loop exits after exhausting the buffer)